src/coreComponents/linearAlgebra/utilities/ReverseCutHillMcKeeOrdering.cpp

```diff
@@ -0,0 +1,330 @@
+                            arrayView1d< localIndex > const marker )
+{
+  localIndex i = 0;
+  localIndex r1 = 0;
+  localIndex r2 = 0;
+  localIndex row = 0;
+  localIndex minDegree = 0;
+  localIndex levDegree = 0;
+  localIndex numLevels = 0;
+  localIndex newNumLevels = 0;
+  localIndex root = rootRef;
+
+  // at most numRows levels
+  array1d< localIndex > level_i( numRows+1 );
+  array1d< localIndex > level_j( numRows );
+
+  // build initial level structure from root
+  buildLevel( offsets, columns, numRows, rankOffset, root,
+              marker, level_i, level_j, newNumLevels );
+
+  numLevels  = newNumLevels - 1;
+  while( numLevels < newNumLevels )
+  {
+    numLevels = newNumLevels;
+    r1 =  level_i[numLevels - 1];
+    r2 =  level_i[numLevels];
+    minDegree = numRows;
+    for( i = r1; i < r2; i++ )
+    {
+      // select the last level, pick min-degree node
+      row = level_j[i];
+      levDegree = offsets[row + 1] - offsets[row];
+      if( minDegree > levDegree )
+      {
+        minDegree = levDegree;
+        root = row;
+      }
+    }
+    buildLevel( offsets, columns, numRows, rankOffset, root,
+                marker, level_i, level_j, newNumLevels );
+  }
+  rootRef = root;
+}
+
+void
+ReverseCutHillMcKeeOrdering::
+  buildLevel( localIndex const * const offsets,
+              globalIndex const * const columns,
+              localIndex const numRows,
+              localIndex const rankOffset,
+              localIndex const root,
+              arrayView1d< localIndex > const marker,
+              arrayView1d< localIndex > const level_i,
+              arrayView1d< localIndex > const level_j,
+              localIndex & numLevelsRef )
+{
+  localIndex i = 0;
+  localIndex j = 0;
+  localIndex l1 = 0;
+  localIndex l2 = 0;
+  localIndex lCurrent = 0;
+  localIndex r1 = 0;
+  localIndex r2 = 0;
+  localIndex row_i = 0;
+  localIndex row_j = 0;
+  localIndex numLevels = 0;
+
+  // set first level first
+  level_i[0] = 0;
+  level_j[0] = root;
+  marker[root] = 0;
+  numLevels = 1;
+  l1 = 0;
+  l2 = 1;
+  lCurrent = l2;
+
+  // explore nbhds of all nodes in current level
+  while( l2 > l1 )
+  {
+    level_i[numLevels++] = l2;
+
+    // loop through last level
+    for( i = l1; i < l2; i++ )
+    {
+      // the node to explore
+      row_i = level_j[i];
+      r1 = offsets[row_i];
+      r2 = offsets[row_i + 1];
+      for( j = r1; j < r2; j++ )
+      {
+        row_j = columns[j] - rankOffset;
+        if( row_j >= 0 && row_j < numRows )
+        {
+          if( marker[row_j] < 0 )
+          {
+            // unmarked row
+            marker[row_j] = 0;
+            level_j[lCurrent++] = row_j;
+          }
+        }
+      }
+    }
+    l1 = l2;
+    l2 = lCurrent;
+  }
+  // after this we always have a "ghost" last level
+  numLevels--;
+
+  // reset marker
+  for( i = 0; i < l2; i++ )
+  {
+    marker[level_j[i]] = -1;
+  }
+
+  numLevelsRef = numLevels;
+}
+
+void
+ReverseCutHillMcKeeOrdering::
```

Comment:
Maybe it makes sense to use `std::sort`?

Reply:
Here I left the `qsort` for now because it seems specialized to handle the presence of the permutation, if I understood correctly. However, your comment made me think that I could use `std::swap` instead of the custom swap that was there before

Reply:
Good! Makes sense!